sopherapps/lilly
# Add SQLAlchemyRepository

## What was Done

- In order to make it easy to create repositories that connect to relational databases using [SQLAlchemy](https://sqlalchemy.org), I created an `SQLAlchemyRepository` subclass of `Repository`
- As I was running tests, I realized that connecting to postgreSQL database using the `SQLAlchemyDataSource` was hanging. 
    - The issue seemed to stem from the way the sessions were being created in a generator method on the `SQLAlchemyDataSource` class. Calling `connect` on the same `SQLAlchemyDataSource` instance would close previously opened sessions even before they completed. I transferred the session to its own `ContextManager` which handles the closing on `__exit__`
- I also added an `.example.env` file to show how to create a `.env` file which is useful in running tests with secrets like `TEST_DATABASE_URL`
- To cater for tests running on PostgreSQL, I added a postgres service in the github actions `ci.yml` file in the `.github/workflows` folder

## How to test

- Run the unit tests for both `SQLAlchemyDataSource` and `SQLAlchemyRepository`

```shell
python -m unittest discover -p test_sqlalchemy.py
```
- To see it in action, follow the instructions on the README to create a new app and run the app and open the docs endpoint in your browser according to the instructions. 
- By default, the app is running on sqlite. To use postgreSQL, set the environment variable `DATABASE_URL` to a postgreSQL database url you can access then run the app as before.

```shell
export DATABASE_URL=postgresql://<user>:<password>@<host>/<database_name>
```
